main/src/com/google/refine/expr/functions/strings/ParseUri.java

```diff
@@ -0,0 +1,89 @@
+package com.google.refine.expr.functions.strings;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.refine.expr.EvalError;
+import com.google.refine.grel.Function;
+import com.google.refine.util.ParsingUtilities;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Parse a URI string into its components and returns a JSON object with the following keys:
+ * scheme, authority, path, query, fragment, host, port.
+ */
+public class ParseUri implements Function {
+
+    @Override
+    public Object call(Properties bindings, Object[] args) {
+        if(args.length == 1 && args[0] instanceof String) {
+            String s = (String)args[0];
+            try {
+                URL url = new URL(s);
+                URI uri = url.toURI();
+                ObjectMapper objectMapper = new ObjectMapper();
+
+                // qp represents the query parameters as a single string in the fragment
+                String fragment = "", qp = "";
+                if(uri.getFragment() != null) { // if there is a fragment
+                    fragment = uri.getFragment(); // get the fragment
+                    if(fragment.contains("?")) { // if there is a query string
+                        // split the fragment into the query string and the fragment
+                        String[] parts = fragment.split("\\?");
+
+                        // get the fragment only and query string
+                        fragment = parts[0]; qp = parts[1];
+                    }
+                }
+
+                // initial
+                Map<String, String> queryParams = new HashMap<>(Map.of());
+                if(qp.length() != 0) {
+                    // get the query parameters as a list of name-value pairs
+                    Arrays.stream(qp.split("&"))
+                            .forEach(pair -> queryParams.put(pair.split("=")[0], pair.split("=")[1]));
+                }
+
+                return ParsingUtilities.mapper.readTree(objectMapper.writeValueAsString(Map.of(
+                        "scheme", uri.getScheme() == null ? "" : uri.getScheme(),
```

Comment:
I am not sure why we should convert `null` to `""` here: JSON can store `null` values too, which do exist in GREL, right?

Reply:
We actually do. 
I faced some issues with an earlier implementation that returned a Map instead of a JSONObject, so I had to replace `null` with `""`. This was just carried on from there. 

I will update accordingly, thanks ☺️